src/main/java/org/jenkinsci/plugins/scriptsecurity/sandbox/whitelists/ProxyWhitelist.java

```diff
@@ -66,7 +68,7 @@
     private final List<EnumeratingWhitelist.FieldSignature> staticFieldSignatures = new ArrayList<EnumeratingWhitelist.FieldSignature>();
 
     /** anything wrapping us, so that we can propagate {@link #reset} calls up the chain */
-    private final Map<ProxyWhitelist,Void> wrappers = new WeakHashMap<ProxyWhitelist,Void>();
+    private final Map<ProxyWhitelist,Void> wrappers = Collections.synchronizedMap(new WeakHashMap<ProxyWhitelist,Void>());
```

Comment:
Seems clearer to revert this and explicitly acquire a monitor (`synchronized (wrappers) {…}`) around every access. Or just use `lock` for this.

Reply:
I thought about this. But if using the `lock`, access to the `wrappers` would be locked for the whole reset and I thought this could have a greater impact than granular synchronization / or using thread safe collection ?

When using `lock` and externalizing [pw.wrappers.put(this, null);](https://github.com/jenkinsci/script-security-plugin/blob/script-security-1.77/src/main/java/org/jenkinsci/plugins/scriptsecurity/sandbox/whitelists/ProxyWhitelist.java#L123) in a private method such as:

```
    private void addWrapper(ProxyWhitelist key, Void value) {
        lock.writeLock().lock();
        try {
            this.wrappers.put(key, value);
        } finally {
            lock.writeLock().unlock();
        }
    }
```

The test passes too. And time difference is negligible, oscillates around 2s, like with the `synchronizedMap` solution. If I change the test settings either increase the number of tasks or threads in the pool, time are equivalent between both solutions.